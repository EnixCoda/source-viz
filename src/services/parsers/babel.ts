export async function prepare() {
  const babelParser = await import("@babel/parser");
  const { default: traverse } = await import("@babel/traverse");
  return function parse(source: string) {
    const ast = babelParser.parse(source, {
      sourceType: "module",
      attachComment: false,
      plugins: [
        "jsx",
        "typescript",
        "classProperties",
        "objectRestSpread",
        "optionalChaining",
        "nullishCoalescingOperator",
        "dynamicImport",
      ],
    });
    const dependencies: [string, boolean][] = [];
    for (const node of ast.program.body) {
      if (node.type === "ImportDeclaration") {
        const { source } = node;
        const { value: dependency } = source;
        dependencies.push([dependency, false]);
      }
    }
    // TODO: find `import()` statements
    const shouldScanDynamicImport = source.includes("import(");
    if (shouldScanDynamicImport) {
      // extract `import()` statements
      traverse(ast, {
        Import: (nodePath) => {
          const { parent } = nodePath;
          if (parent.type === "CallExpression") {
            const importTargetNode = parent.arguments[0];
            if (importTargetNode.type === "StringLiteral") {
              const dependency = importTargetNode.value;
              dependencies.push([dependency, true]);
            } else if (importTargetNode.type === "TemplateLiteral") {
              for (const node of importTargetNode.quasis) {
                const dependency = node.value.cooked || node.value.raw;
                dependencies.push([dependency, true]);
              }
            } else {
              throw new Error(`${importTargetNode.type} is not a string`);
            }
          } else {
            throw new Error(`Unexpected parent type "${parent.type}"`);
          }
        },
      });
      // parse extracted statements
      // extract dependencies
    }
    return dependencies;
  };
}
