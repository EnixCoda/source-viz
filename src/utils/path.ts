import { join } from "path-browserify";

export const resolvePath = (...paths: string[]) => {
  const path = join(...paths);
  if (path.match(/\.\./)) throw new Error(`Path "${path}" is lower than root`);
  return path;
};
// Relative:
//    ./
//    ../
// Absolute:
//    path
//    path/to/file
//    /path/to/file
// do not use `path.isAbsolute` here, because of usages like `lib/path`

export const isRelativePath = (path: string) => path.startsWith(".");
