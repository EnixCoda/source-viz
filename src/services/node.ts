import * as fs from "fs/promises";
import path from "path";

const resolveSymlink = false;

export async function getFiles(
  rootDir: string,
  isExcluded: (path: string) => boolean
) {
  const files: string[] = [];

  async function scan(relativePath: string) {
    if (isExcluded(relativePath)) return;

    const absolutePath = path.resolve(rootDir, relativePath);
    const stats = await fs.stat(absolutePath);
    if (stats.isDirectory()) {
      const items = await fs.readdir(absolutePath);
      for (const item of items) await scan(path.join(relativePath, item));
    } else if (stats.isFile()) {
      files.push(relativePath);
    } else if (stats.isSymbolicLink()) {
      if (resolveSymlink) {
        // TODO
        console.info(`Skipped symbolic link "${relativePath}"`);
      }
    } else {
      console.error(`Unrecognized item type "${relativePath}"`);
    }
  }

  await scan(".");

  return files;
}
