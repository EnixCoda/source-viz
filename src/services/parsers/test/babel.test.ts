// test parsing './import-and-requires' with babel
import { promises as fs } from "node:fs";
import * as path from "node:path";
import * as babelParser from "../babel";

describe("babel", () => {
  it("should parse import and require statements", async () => {
    const fileContent = await fs.readFile(path.join(__dirname, "import-and-require.js"), "utf-8");
    const parse = await babelParser.prepare();
    const ast = parse(fileContent);
    expect(ast).toMatchSnapshot();
  });
});
