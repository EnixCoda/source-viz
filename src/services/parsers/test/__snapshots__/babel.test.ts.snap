// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel should parse import and require statements 1`] = `
[
  [
    "b0.js",
    false,
  ],
  [
    "b1.js",
    false,
  ],
  [
    "b2.js",
    false,
  ],
  [
    "b3.js",
    true,
  ],
  [
    "b4.js",
    true,
  ],
  [
    "b9.js",
    false,
  ],
  [
    "b10.js",
    false,
  ],
  [
    "b14.js",
    false,
  ],
  [
    "b15.js",
    true,
  ],
]
`;
